src/presenter/films-presenter.js

```diff
@@ -1,102 +1,150 @@
 import {render, remove} from '../framework/render.js';
-import {SHOW_FILM_COUNT_STEP} from '../const.js';
+import {SHOW_FILM_COUNT_STEP, SortType, UpdateType, UserAction, FilterType} from '../const.js';
 import FilmSectionView from '../view/film-section-view.js';
 import FilmContainerView from '../view/film-container-view.js';
 import LoadMoreButtonView from '../view/load-more-button-view.js';
 import NoFilmCardView from '../view/no-film-card-view.js';
 import SortView from '../view/sort-view.js';
 import FilmPresenter from './film-presenter.js';
-import {updateItem} from '../utils/common.js';
 import {sortFilmsByRating, sortFilmsByDate} from '../utils/task.js';
-import {SortType} from '../const.js';
+import {filter} from '../utils/filter.js';
 
 export default class FilmsPresenter {
 
   #filmSection = new FilmSectionView;
   #filmContainer = new FilmContainerView;
-  #loadMoreButtonComponent = new LoadMoreButtonView();
-  #noFilmComponent = new NoFilmCardView();
-  #sortComponent = new SortView();
+  #loadMoreButtonComponent = null;
+  #noFilmComponent = null;
+  #sortComponent = null;
 
-  #sectionMovie = [];
   #movieModel = null;
   #filmListContainer = null;
   #renderedMovieCount = SHOW_FILM_COUNT_STEP;
 
   #filmPresenter = new Map();
   #currentSortType = SortType.DEFAULT;
-  #sourcedFilms = [];
+  #filterType = FilterType.ALL;
 
-  constructor(filmListContainer, movieModel) {
+  #filterModel = null;
+
+  constructor(filmListContainer, movieModel, filterModel) {
     this.#filmListContainer = filmListContainer;
     this.#movieModel = movieModel;
+    this.#filterModel = filterModel;
+
+    this.#movieModel.addObserver(this.#handleModelEvent);
+    this.#filterModel.addObserver(this.#handleModelEvent);
+  }
+
+  get movies() {
+    this.#filterType = this.#filterModel.filter;
+    const movies = this.#movieModel.movies;
+    const filteredMovies = filter[this.#filterType](movies);
+
+    switch (this.#currentSortType) {
+      case SortType.RATING:
+        return filteredMovies.sort(sortFilmsByRating);
+      case SortType.DATE:
+        return filteredMovies.sort(sortFilmsByDate);
+    }
+
+    return filteredMovies;
   }
 
   init = () => {
-    this.#sectionMovie = [...this.#movieModel.movie];
-    this.#sourcedFilms = [...this.#movieModel.movie];
     this.#renderMovie();
   };
 
```

Comment:
Добавим метод для обновления модального окна после перерисовки
```suggestion
  #updateOpenedModal = () => {
    if (!this.#openedFilmPresenter) {
      return;
    }

    if (!this.#openedFilmPresenter.isOpened) {
      this.#openedFilmPresenter = null;
    }
    const currentModalData = this.#movieModel.movies.find((movie) => movie.id === this.#openedFilmPresenter.movieId);
    if (currentModalData) {
      this.#openedFilmPresenter.init(currentModalData);
    }
  };
```